### PR TITLE
disable use_cache if using gradient checkpointing

### LINKED ISSUE #30301: Idefics2 raises shape error in the backward pass with gradient_checkpointing
I get a shape error if I PEFT the latest Idefics2 model at head with gradient checkpointing. The symptom is identical to [this](https://stackoverflow.com/questions/77656467/attention-mask-error-when-fine-tuning-mistral-7b-using-transformers-trainer). Seems like the following block isn't compatible with gradient checkpointing.
https://github.com/huggingface/transformers/blob/e4ea19b958c89d61e42461fac6ac8441787121f8/src/transformers/models/idefics2/modeling_idefics2.py#L1593-L1596

In my local workspace a fix similar to https://github.com/huggingface/transformers/pull/28061/commits/91f4b7e2b8c305dfa9044463833ecbb0ae6ba560 removed the shape error but a quick review of the logic from a maintainer would be lovely. Happy to submit PR after. Thanks!

### PR DESCRIPTION
**Pull request by chenzizhao**
# What does this PR do?

I disable `use_cache` when gradient checkpointing is enabled in the new Idefics2 (like in Idefics, Mistral7B, and many others). Previously, the backward pass raises a shape error.

Fixes #30301

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

cc: @amyeroberts

### CODE REVIEW

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this fix so quickly!
